PublicMapping/district-builder-2
# Support producing a container image and serving static assets

## Overview

- Parameterize TypeORM connection info, via the environment
- Remove TypeORM config object from `app.module.ts`, to ensure there is a single source of truth
- Configure React proxy via `http-proxy-middleware`, to support parametrization
- Serve static files with [`@nestjs/serve-static`](https://github.com/nestjs/serve-static)
- Refactor `update`, `test`, and `cibuild` to support pipeline that produces a single container image with the frontend bundle and the Nest backend
- Configure CI with Jenkins
- Refactor project README

Resolves #19 
Resolves #20 

## Demo

<img width="1745" alt="image" src="https://user-images.githubusercontent.com/1774125/74660822-e0e67300-5164-11ea-91eb-24ea40d9b236.png">

## Testing Instructions

- Refresh the development environment
```bash
$ docker-compose down ; ./scripts/update
```
- Ensure that tests pass
```bash
$ ./scripts/test
```
- See that the backend is able to serve static files
```bash
$ ./scripts/server
. . .
$ http --headers localhost:3005
HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: public, max-age=60
Connection: keep-alive
Content-Length: 2281
Content-Type: text/html; charset=UTF-8
Date: Mon, 17 Feb 2020 14:03:46 GMT
ETag: W/"8e9-1705370e71b"
Last-Modified: Mon, 17 Feb 2020 13:57:53 GMT
X-Powered-By: Express
```
- Compare `Cache-Control` headers with that of [WhiteNoise](http://whitenoise.evans.io/en/stable/django.html):
```
$ http --headers --verify=no https://[OAR production ALB]/static/media/OpenApparelRegistry_logo.11bd8a29.png Host:openapparel.org
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Cache-Control: public, max-age=60
Connection: keep-alive
Content-Length: 74556
Content-Type: image/png
Date: Mon, 17 Feb 2020 14:04:35 GMT
Last-Modified: Wed, 12 Feb 2020 18:38:52 GMT
Server: gunicorn/20.0.4
X-Frame-Options: SAMEORIGIN
```
- See that we are able to execute migrations using the single container image (this models how we'd invoke a management command on ECS):
```bash
[rocky@algiers ~]$ docker run \
    --network district-builder-2_default \
    -e POSTGRES_HOST=database \
    -e POSTGRES_USER=districtbuilder \
    -e POSTGRES_PASSWORD=districtbuilder \
    -e POSTGRES_DB=districtbuilder \
    -it --entrypoint bash district-builder-server:latest -c "yarn migration:run"
yarn run v1.22.0
$ yarn run typeorm migration:run
$ ts-node -r tsconfig-paths/register ./node_modules/typeorm/cli.js migration:run
Couldn't find tsconfig.json. tsconfig-paths will be skipped
query: SELECT * FROM "information_schema"."tables" WHERE "table_schema" = current_schema() AND "table_name" = 'migrations'
query: SELECT * FROM "migrations" "migrations" ORDER BY "id" DESC
No migrations are pending
Done in 1.31s.
```
